source/mongodb/tables/role-configuration.rst

```diff
@@ -36,82 +41,149 @@
 
    * - | ``apply_when``
        | Expression
-     - An :ref:`expression <expressions>` that evaluates to ``true`` when
-       this role applies to a user for a specific document.
+     - An :ref:`expression <expressions>` that evaluates to true when
+       this :ref:`role <roles>` applies to a user for a specific document.
+
+       For a given request on a document, App Services evaluates the
+       ``apply_when`` of each relevant role until one evaluates to true.
+       Relevant roles are any roles specified for the document's collection plus
+       the default role(s). For more information, see :ref:`roles`.
+
+       If Device Sync (Flexible Mode) is enabled, the role must be :ref:`sync
+       compatible <sync-compatible-roles>`. If the role is not sync compatible,
+       but its ``apply_when`` evaluated to true, others roles are not
+       considered; access is denied.
+
+   * - | ``document_filters``
+       | Document
+       | *Default:* ``undefined``
+     - A document with read and write expressions that determine whether
+       the role's other permissions may be evaluated.
+       
+       If undefined, and Device Sync is not enabled, the role's other
+       permissions may be evaluated. If Device Sync is enabled, both
+       ``document_filters.read`` and ``document_filters.write`` **must** be
+       defined for this role to be :ref:`sync compatible
+       <sync-compatible-roles>`.
```

Comment:
That looks right to me, I think there are probably two more columns worth expanding that table to though:

| | document_filters is undefined | document_filters.read is undefined, document_filters.write is defined | document_filters.read is defined, document_filters.write is undefined | document_filters.read and document_filters.write are both defined |
| - | -  | - | - | - |
| flex sync enabled | sync incompatible (deny access)  | sync incompatible (deny access) | sync incompatible (deny access) | if `document_filters.write` evaluates to true, evaluate read/write perms. if `document_filters.write` evaluates to false, deny write access, and evaluate `document_filters.read` - if `document_filters.read` evaluates to true, evaluate read perms |
| flex sync disabled | evaluate read/write perms | if `document_filters.write` evaluates to true, evaluate read/write perms; otherwise, deny read/write access | if `document_filters.read` evaluates to true, evaluate read perms; deny writes  | same as above |  

By `evaluate read/write perms` in the table above, I mean look at top-level `read`/ top-level `write`, and the corresponding field-level perms if top-level `read`/`write` are omitted (ie, the existing role permission evaluation)

Reply:
I think the diagrams that Caleb has been working on will help communicate this better!

Reply:
Tried reworking, have a look (when uploaded)